### DISCUSSION TITLE
AD support with home directory creation

### DISCUSSION THREAD

**Discussion by carles-alemany** [ltsp/ltsp#408] [General]
Hi everyone, and thank you for this great project!

We've been using LTSP5 over the last 8 years, and now it's time to upgrade. On our old setup, which is still running, we have a Samba4 AD VM + LTSP5 VM serving different chroots to fat clients. LTSP server is joined to AD via pbis-open, fat-clients are able to authenticate with domain credentials and, every time a new user logs in, a new home directory is created into the LTSP server.

We tried to replicate this setup with the new LTSP version, working with chroots as we did before. At this point, everything works, users are able to login with AD credentials but home folders are being created on the clients, not on the server, so everytime a client is disconnected, user's home directory is lost.

What we've done:

Create a new chroot and assign a password to root.
Export the new created chroot via NFS in order to boot from a client (https://github.com/ltsp/ltsp/wiki/chroots “Booting the chroot with a client”).
Boot a client using the exported chroot.
Join the chroot to AD with realmd-join, following these steps: https://computingforgeeks.com/join-ubuntu-debian-to-active-directory-ad-domain/
After joining the chroot to AD, its name (focal) is being shown as a new computer inside the AD.
Shutdown the client
Update chroot on the server (sudo update ltsp focal)
After doing these steps, a client can boot from the newly created image and we're able to login with a domain user and access SMB shares. Everything works except home directory, which is being created at the client instead of the server.

Also, during the login, the client shows this message but then logs in without issues:
![108489844-56b67a80-72a2-11eb-8ca3-ba48c8de0850](https://user-images.githubusercontent.com/79309152/109282870-65f77400-781e-11eb-94be-58900ed74e0c.png)

Any ideas on how to automatically create user's home directory on the server at login? I'll be more than happy to make a wiki page if I can make it work.

Thank you!

**Comment by alkisg** [comment]
Hello, what happens if you login as root on the client, and then you type this command?

```shell
ssh username@server
```

In the place of `username`, put an LDAP user.
Do not replace `server` with a real hostname, because the LTSP clients know the LTSP server with that codename, "server".

Are you able to SSH to the server that way?
If yes, can you install and configure pam_mkhomedir on the server, so that the home directory of SSH users is automatically created?

Finally, what is the output of the following command, both on the server and on a client?

```shell
getent passwd username
```

**Comment by carles-alemany** [comment]
Hello, 

I've tried to login as root on the client, and I get this message: "User root can't auth via pamltsp"

So, I logged in as a user with sudo permissions on the client, and then I did:

`sudo su`

Then, as root:

`root@ltsp249:/home/adminlocal# ssh profetest2@domain.lan@server`

After putting the password, this message appears:

`Permission denied, please try again.`

So, I'm unable to login via ssh to the server.

Here are the outputs you asked for: 

**SERVER**:
The command does not give any output at all:
`adminlocal@S-208:~$ getent passwd profetest2@domain.lan`

**CLIENT**:
`adminlocal@ltsp249:~$ getent passwd profetest2@domain.lan
profetest2@domain.lan:*:1455407248:1455400513:Profetest 2:/home/profetest2@domain.lan:/bin/bash`

Maybe I should join the server to the domain too, doing the same steps I did with the chroot?

Thank you

**Comment by alkisg** [comment]
To log in as root on a client, follow the last example of https://ltsp.org/man/ltsp.conf/.
You don't need to, as you already have a way to do it.

In your old LTSP5 setup, you should have been able to run `ssh user@server`, otherwise SSHFS wouldn't work. If that's not the case now, it means you omitted something while setting up the system.

So yeah either join the server to AD, or switch to NFS home instead of SSHFS home.

I'll move this issue to discussions as it's not about a bug in the LTSP code.

**Reply by carles-alemany** [reply]
Hello,

Now I've joined the server into the domain following [this](https://computingforgeeks.com/join-ubuntu-debian-to-active-directory-ad-domain/) guide, which also includes setting up pam_mkhomedir.

Now, if I login to a root terminal on a client and I do `ssh user@domain@server`, it asks me the user's password and I can login to the server with the domain user. 

Unfortunately, the behaviour with home dirs is the same. If I login with a domain user directly on the server, it creates a permanent /home dir. If i login from a client (chroot image), it creates a temporary /home, which is being destroyed when I log out. 

Both the server and the chroot are joined to the domain separately, as independent machines. With LTSP5, I only had to join the server to make it work. 

I appreciate your help.

**Comment by alkisg** [comment]
> With LTSP5, I only had to join the server to make it work.

That was a huge security issue and it's no longer an option.

Run the following command on a client as root. What is the result?

```shell
sshfs user@domain@server: /mnt
ls /mnt
```

**Reply by carles-alemany** [reply]
Hi, 

The command does not give any output at all: 

```
root@ltsp249:~# sshfs profetest3@domain.lan@server: /mnt
profetest3@domain.lan@server's password: 
root@ltsp249:~# ls /mnt
```

Thank you

**Comment by alkisg** [comment]
This means that sshfs completed successfully and you just don't have any visible files in /home/profetest.

As root again, what happens if you run these? Does home get mounted? Do any error messages appear?

```shell
umount /mnt
login profetest3
mount | grep home
```

**Reply by carles-alemany** [reply]
Here is the output:

```
root@ltsp249:~# umount /mnt
root@ltsp249:~# login profetest3@domain.lan
Password: 
Welcome to Ubuntu 20.04 LTS (GNU/Linux 5.4.0-26-generic x86_64)

The programs included with the Ubuntu system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Ubuntu comes with ABSOLUTELY NO WARRANTY, to the extent permitted by
applicable law.


The programs included with the Ubuntu system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Ubuntu comes with ABSOLUTELY NO WARRANTY, to the extent permitted by
applicable law.

Creating directory '/home/profetest3@domain.lan'.
Invalid passwd/shadow for profetest3@domain.lan
profetest3@domain.lan@ltsp249:~$ mount | grep home
adminlocal@server: on /home/adminlocal type fuse.sshfs (rw,nosuid,nodev,relatime,user_id=0,group_id=1455400513,default_permissions,allow_other)
```

"adminlocal" is a local (outside the domain) user created both at the server and inside the chroot. This user is inside sudoers group.

When I login to the client as "adminlocal", the home at the client is the one stored at the server. The issue appears only when I login with domain users. 

Thank you

**Comment by alkisg** [comment]
Run the following commands as root on the client, and upload their output:

```shell
grep profetest3 /etc/passwd
grep profetest3 /etc/shadow
getent passwd profetest3
getent shadow profetest3
````

If you see any password hashes in the output of the commands (you shouldn't), then just replace them with XXX or something.

**Reply by carles-alemany** [reply]
Hi, 

Here's the output:

```
root@ltsp249:~# grep profetest3@domain.lan /etc/passwd
root@ltsp249:~# grep profetest3@domain.lan /etc/shadow
root@ltsp249:~# getent passwd profetest3@domain.lan
profetest3@domain.lan:*:1455407318:1455400513:Profetest 3:/home/profetest3@domain.lan:/bin/bash
root@ltsp249:~# getent shadow profetest3@domain.lan

```

Thank you

**Comment by alkisg** [comment]
Hrm, based on that output, I can't see what's wrong.

So. On the client, open this file:

```shell
sudo nano /usr/share/ltsp/client/login/pamltsp
```

Locate the pam_auth() function. At its beginning, add the following two lines:

```shell
pam_auth() {
exec 2 >> /tmp/log.txt
set -x
# ...the rest of the function follow here...
```

Then try once more to `login profetest3@domain.lan`. It'll fail. Then, upload the contents of /tmp/log.txt, which will be a command trace. Do check if any sensitive info is there, if so remove it.

**Comment by carles-alemany** [comment]
Hi, 

I've added the two lines to the function and did the login. Here is the output:

```
root@ltsp249:/home/adminlocal# login profetest3@domain.lan
Password: 
Welcome to Ubuntu 20.04 LTS (GNU/Linux 5.4.0-26-generic x86_64)

The programs included with the Ubuntu system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Ubuntu comes with ABSOLUTELY NO WARRANTY, to the extent permitted by
applicable law.


The programs included with the Ubuntu system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Ubuntu comes with ABSOLUTELY NO WARRANTY, to the extent permitted by
applicable law.

Creating directory '/home/profetest3@domain.lan'.
/usr/share/ltsp/client/login/pamltsp: 52: exec: 2: not found
profetest3@domain.lan@ltsp249:~$ exit
```


The file log.txt was empty. I tried gaving it 777 permissions and login in again, but the file still was empty. 

I'm doing those tests on an independent environment, so if necessary I can reinstall everything from scratch and detail all the steps. Maybe it can be helpful.

Thank you

**Comment by alkisg** [comment]
I meant for you to edit the pamltsp script, not to run the commands in a terminal.

If it's not a production environment, come over to IRC some time so that I can have a look over screen sharing. I.e. go to ltsp.org and click the IRC link.

**Reply by carles-alemany** [reply]
I edited the pamltsp script with nano, then I did the login on a terminal using root. Coming to IRC!

**Comment by alkisg** [comment]
For reference, the problem was just an older LTSP version where LDAP wasn't supported yet.

The chroot was created using an old LTSP, then `ltsp initrd` put the old version in ltsp.img, then the chroot was updated using the PPA but the old version was still in ltsp.img until we ran `ltsp initrd` once more.

**Comment by danboid** [comment]
Hi @carles-alemany 

We are in the process of replacing our ancient LTSP5 16.04 server with an updated one but the area I'm most unsure about is the auth.

We are currently using the openldap client to provide LDAP auth using a Windows 2016 DC under chrootless LTSP5.

Seeing as most of the machines here run Windows, it would make sense for us to use AD for auth, provided its not too much extra effort but configuring LTSP to use sssd-ad sounds like a LOT of extra effort according to this guide:

https://github.com/ltsp/ltsp/discussions/516

That guide says every LTSP client needs to be registered with the DC, which would be  a lot of work for us and would be a deal breaker but it sounds like you only had to register one chroot with your DC (using realmd) and previously you had LTSP AD auth working using pbis-open? Does pbis-open not work with the latest LTSP / Ubuntu 20.04? Have you got the latest LTSP working fine now with AD and Ubuntu 20.04?

There seems to be a lot of uncertainty and complication in switching to using AD for LTSP so I think we'll be sticking with LDAP but even then I'm not sure if we continue using our currently openldap config (if we can still use it) or switch to sssd-ldap. I understand sssd-ldap would enable caching of logins but I'm not sure how much extra config is required versus using openldap and if its worth the effort?

**Reply by maccadia** [reply]
The whole point of joining a computer to AD is to establish a trust between the AD and the client. Since LTSP clients are created "on the fly" I'm not sure it makes sense to joins clients to AD. (There is maybe a way to join / leave AD at boot and shutdown but it's not simple).

Using LDAP with sssd is the best option for you I think. I would say it's not that complicated to set up. Be aware though that your LDAP server should be reachable directly from your LTSP clients (in LTSP5 it was only needed for the LTSP server). SSSD is only client side so you can keep your OpenLDAP config server side (maybe with small changes).

**Reply by carles-alemany** [reply]
Hi @danboid,

What I did was to export the chroot through NFS, boot it using a client and joining it to AD domain using SSSD. Here are the steps I've followed:

1. Export chroot using NFS: https://github.com/ltsp/ltsp/wiki/chroots#chroots
2. Boot the chroot using a client (https://github.com/ltsp/ltsp/wiki/chroots#booting-the-chroot-with-a-client).
3. On the client, join it to AD following those steps: https://computingforgeeks.com/join-ubuntu-debian-to-active-directory-ad-domain/
4. Shut down the client
5. On the server, do:
`
sudo ltsp image your_chroot_name
sudo ltsp initrd
`
6. Disable the NFS share we created, in order to prevent other clients to boot from it.
7. Run `ltsp ipxe` in order to remove the entry "nfs-rw" from the clients' boot menu.

I hope this works for you.

**Reply by danboid** [reply]
Hi @carles-alemany 

Thanks for tagging me in this. We are still using LDAP but want to move to LDAPS or AD for LTSP auth purposes ASAP for security reasons.

It seems you fixed your problem with the home dirs not getting created on the server correct? Do you still get the error on the login screen?

If your method works without any issue then we should add a new page to the LTSP wiki, which at the moment only offers this, unusable for most, guide:

https://github.com/ltsp/ltsp/wiki/LTSP-and-Samba-Windows-NT-Domain

**Reply by carles-alemany** [reply]
Hi @danboid,

This config works without any issue, almost one year in production at the moment.

The error on the login screen is gone, it was due to an ltsp version mismatch between the server and the chroot, @alkisg helped me to solve it. After updating both of them and doing `ltsp initrd`, everything started working.

Regarding the config I used, despite it works and it's stable, I see some drawbacks:

- It has not been tested in a chrootless environment, only with chroot.
- It works only for login, you cannot connect to a CIFS share on the same domain without inputting your password as you would do in a standalone machine joined to the domain.

You can solve the CIFS problem using a simple login script that asks the user for their credentials and then mounts the share, or simply by telling users that they'll have to click the share shortcut when needed and input their password. 

With a login script you could even encrypt and store the user's password somehow at their /home to avoid asking the password twice at every login, but it can be a security risk depending on the environment, because the script has to be able to decrypt the password, so it's not safe at all.

I agree with creating the wiki page, I can re-create my testing environment and document all the steps, and then you can test and see if it works for you too.

**Reply by danboid** [reply]
We're using VM images for LTSP so hopefully this method will work for those too. We don't use CIFS shares under LTSP so that shouldn't be an issue.

Please let me know if and when you document your steps on the wiki. I'm sure we can't be the only other LTSP users who would like to use it with AD.
